app/services/invoices/finalize_batch_service.rb

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+module Invoices
+  class FinalizeBatchService < BaseService
+    def initialize(organization:)
+      @organization = organization
+
+      super
+    end
+
+    def call_async
+      Invoices::FinalizeAllJob.perform_later(organization:, invoice_ids: invoices.ids)
+
+      result.invoices = invoices
+
+      result
+    end
+
+    def call(invoice_ids)
+      processed_invoices = []
+      Invoice.where(id: invoice_ids).find_each do |invoice|
+        result = Invoices::RefreshDraftAndFinalizeService.new(invoice:).call
+
+        return result unless result.success?
```

Comment:
I think we should continue to process all the invoices and not fail after the first error.
A buggy invoice would prevent the batch to ever finish.

I'd return 2 list of invoices in the final result and set error if there was any error

```rb
result.processed_invoices = processed_invoices.compact
result.failed_invoices = failed.compact

result.fail_with_error!(error: :something)

result
```

Reply:
that's a tricky one, because if we failed with an integration, possible that it means that the integration is not set up correctly, and therefore we'll fail with all further fetches as well...
And anrok is slow in responding + they have their limits... 🤔  so in this case it might be better to be sure that the error is fixed before continuing